API/Helpers/PdfMetadataExtractor.cs

```diff
@@ -0,0 +1,1605 @@
+                PutBack();
+
+                break;
+            }
+        }
+
+        switch (sb.ToString())
+        {
+            case "true":
+                return new Token(TokenType.Bool, true);
+
+            case "false":
+                return new Token(TokenType.Bool, false);
+
+            case "stream":
+                return new Token(TokenType.StreamStart, true);
+
+            case "endstream":
+                return new Token(TokenType.StreamEnd, true);
+
+            case "endobj":
+                return new Token(TokenType.ObjectEnd, true);
+
+            default:
+                return new Token(TokenType.Keyword, sb.ToString());
+        }
+    }
+}
+
+class PdfMetadataExtractor : IPdfMetadataExtractor
+{
+    public PdfMetadataExtractor(ILogger<BookService> logger, string filename)
+    {
+        Stopwatch timing = Stopwatch.StartNew();
+        _logger = logger;
+        _stream = File.OpenRead(filename);
+        _lexer = new PdfLexer(_stream);
+
+        ReadObjectOffsets();
+        long objOffsetTime = timing.ElapsedMilliseconds;
+
+        ReadMetadata(filename);
+        long metadataTime = timing.ElapsedMilliseconds - objOffsetTime;
+
+        LogMetadata(filename);
+        timing.Stop();
+
+        _logger.LogInformation("PDF {File}, object offsets {ObjOffsetTime} ms, metadata {MetadataTime} ms", filename, objOffsetTime, metadataTime);
+    }
+
+    public Dictionary<string, string> GetMetadata()
+    {
+        return _metadata;
+    }
+
+    private readonly ILogger<BookService> _logger;
+    private readonly PdfLexer _lexer;
+    private readonly FileStream _stream;
+    private long[] _objectOffsets = new long[0];
+    private readonly Dictionary<string, string> _metadata = new();
+
+    private struct MetadataRef
+    {
+        public long root;
+        public long info;
+
+        public MetadataRef(long root, long info)
+        {
+            this.root = root;
+            this.info = info;
+        }
+    }
+
+    private readonly Stack<MetadataRef> metadataRef = new();
+
+    private void LogMetadata(string filename)
+    {
+        _logger.LogDebug("Metadata for {Path}:", filename);
+
+        foreach (var entry in _metadata)
+        {
+            _logger.LogDebug("   {Key:0,-5} : {Value:1}", entry.Key, entry.Value);
+        }
+    }
+
+    private void ReadObjectOffsets()
+    {
+        _stream.Seek(-32, SeekOrigin.End);
+
+        long xrefOffset = _lexer.GetXRefStart();
+
+        ReadXRefAndTrailer(xrefOffset);
+    }
+
+    private void ReadXRefAndTrailer(long xrefOffset)
+    {
+        _stream.Seek(xrefOffset, SeekOrigin.Begin);
+        _lexer.ResetBuffer();
+
+        if (!_lexer.TestByte((byte)'x'))
+        {
+            ReadXRefStream();
+
+            return;
+        }
+
+        var token = _lexer.NextToken();
+
+        if (token.type != PdfLexer.TokenType.Keyword || (string)token.value != "xref")
+        {
+            throw new Exception("Expected xref keyword");
```

Comment:
Since you asked about this, we could use a PdfMetadataExtractorException or something to capture everything coming out of here.